core/src/autogluon/core/hpo/executors.py

```diff
@@ -77,18 +87,121 @@ def register_resources(self, initialized_model: AbstractModel):
-            model_default_num_cpus=resource_kwargs.get('num_cpus', 0),
+        minimum_model_resources = initialized_model.get_minimum_resources(
+            is_gpu_available=(num_gpus > 0)
         )
+        minimum_model_num_cpus = minimum_model_resources.get('num_cpus', 1)
+        minimum_model_num_gpus = minimum_model_resources.get('num_gpus', 0)
+        initialized_model_params = initialized_model.get_params()
+
+        if 'hyperparameters' in initialized_model_params and 'ag_args_fit' in initialized_model_params['hyperparameters']:
+            user_specified_trial_num_cpus = initialized_model_params['hyperparameters']['ag_args_fit'].get('num_cpus', None)
+            user_specified_trial_num_gpus = initialized_model_params['hyperparameters']['ag_args_fit'].get('num_gpus', None)
+            if user_specified_trial_num_cpus is not None or user_specified_trial_num_gpus is not None:
+                num_trials_in_parallel_with_gpu = math.inf
+                if user_specified_trial_num_cpus is None:
+                    # If user didn't specify cpu per trial, we find the min based on gpu
+                    num_trials_in_parallel_with_gpu = num_gpus // user_specified_trial_num_gpus
+                    user_specified_trial_num_cpus = num_cpus // num_trials_in_parallel_with_gpu  # keep gpus per trial int to avoid complexity
+                num_trials_in_parallel_with_cpu = math.inf
+                if user_specified_trial_num_gpus is None:
+                    # If user didn't specify gpu per trial, we find the min based on cpu
+                    num_trials_in_parallel_with_cpu = num_cpus // user_specified_trial_num_cpus
+                    user_specified_trial_num_gpus = num_gpus // num_trials_in_parallel_with_cpu  # keep gpus per trial int to avoid complexity
+                assert user_specified_trial_num_cpus <= num_cpus, \
+                    f"Detected trial level cpu requirement = {user_specified_trial_num_cpus} > total cpu granted to AG predictor = {num_cpus}"
+                assert user_specified_trial_num_cpus >= minimum_model_num_cpus, \
+                    f"The trial requires minimum cpu {minimum_model_num_cpus}, but you only specified {user_specified_trial_num_cpus}"
+                assert user_specified_trial_num_gpus <= num_gpus, \
+                    f"Detected trial level gpu requirement = {user_specified_trial_num_gpus} > total gpu granted to AG predictor = {num_gpus}"
+                assert user_specified_trial_num_gpus >= minimum_model_num_gpus, \
+                    f"The trial requires minimum gpu {minimum_model_num_gpus}, but you only specified {user_specified_trial_num_gpus}"
+
+                # Custom backend should set its total resource to be resources_per_trial
+                self.hyperparameter_tune_kwargs['resources_per_trial'] = {
+                    'num_cpus': user_specified_trial_num_cpus,
+                    'num_gpus': user_specified_trial_num_gpus
+                }
```

Comment:
Unsure where self.hyperparameter_tune_kwargs is defined, but just be careful to not be editing outer context. Maybe this is ok, depends on where it is coming from.

Reply:
Good point. Will double check